be/src/storage/persistent_index.cpp

```diff
@@ -922,11 +1354,561 @@ StatusOr<std::unique_ptr<MutableIndex>> MutableIndex::create(size_t key_size, co
+    const auto avg_num_entry = num_entry / _key_shard_offset_and_sizes.size();
+    // vector<std::pair<key size, <key-value references>>>
+    std::vector<std::pair<uint32_t, std::vector<std::vector<KVRef>>>> kv_refs(_key_shard_offset_and_sizes.size());
+    for (const auto key_shard_offset_and_size : _key_shard_offset_and_sizes) {
+        auto [shard_offset, shard_size] = key_shard_offset_and_size.second;
+        kv_refs.push_back(
+                std::make_pair(key_shard_offset_and_size.first,
+                               _shards[shard_offset]->get_kv_refs_by_shard(shard_size, avg_num_entry, without_null)));
+    }
+    return kv_refs;
+}
+
+bool ShardByLengthMutableIndex::load_snapshot(phmap::BinaryInputArchive& ar_in) {
+    for (const auto& shard : _shards) {
+        if (!shard->load_snapshot(ar_in)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+size_t ShardByLengthMutableIndex::dump_bound() {
+    int size = 0;
+    for (const auto& shard : _shards) {
+        size += shard->dump_bound();
+    }
+    return size;
+}
+
+bool ShardByLengthMutableIndex::dump(phmap::BinaryOutputArchive& ar_out) {
+    for (const auto& shard : _shards) {
+        if (!shard->dump(ar_out)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+Status ShardByLengthMutableIndex::commit(MutableIndexMetaPB* meta, const EditVersion& version, const CommitType& type) {
+    std::shared_ptr<FileSystem> fs;
+    ASSIGN_OR_RETURN(fs, FileSystem::CreateSharedFromString(_path));
+    switch (type) {
+    case kFlush: {
+        // create a new empty _l0 file because all data in _l0 has write into _l1 files
+        std::string file_name = get_l0_index_file_name(_path, version);
+        WritableFileOptions wblock_opts;
+        wblock_opts.mode = FileSystem::CREATE_OR_OPEN_WITH_TRUNCATE;
+        ASSIGN_OR_RETURN(auto wfile, fs->new_writable_file(wblock_opts, file_name));
+        DeferOp close_block([&wfile] {
+            if (wfile) {
+                wfile->close();
+            }
+        });
+        meta->clear_wals();
+        IndexSnapshotMetaPB* snapshot = meta->mutable_snapshot();
+        version.to_pb(snapshot->mutable_version());
+        PagePointerPB* data = snapshot->mutable_data();
+        // create a new empty _l0 file, set _offset to 0
+        data->set_offset(0);
+        data->set_size(0);
+        meta->set_format_version(PERSISTENT_INDEX_VERSION_2);
+        _offset = 0;
+        std::fill(_page_sizes.begin(), _page_sizes.end(), 0);
+        break;
+    }
+    case kSnapshot: {
+        std::string file_name = get_l0_index_file_name(_path, version);
+        // be maybe crash after create index file during last commit
+        // so we delete expired index file first to make sure no garbage left
+        FileSystem::Default()->delete_file(file_name);
+        size_t snapshot_size = dump_bound();
+        phmap::BinaryOutputArchive ar_out(file_name.data());
+        if (!dump(ar_out)) {
```

Comment:
I think a snapshot of `ShardByLengthMutableIndex` is a  snapshot set of meta, and we should store the snapshot set continuously in L0 file.

Reply:
already does so, this dump is this->dump, not MutableIndex's dump.